### PR TITLE
Return revert commit from `Branch.revert()` instead of None

### PR DESCRIPTION
**Pull request by nicholasjng**
This makes it easier to analyze the new state of the repository after the revert. Obtains the commit by querying `Branch.head` after the revert API request, which incurs another branch API call.

Said `Branch.head` call cannot incur any error that wouldn't already fly on the revert attempt, so this does not need any more error handling.

Also allows passing `ReferenceType`s to Branch.revert() to get it up to par with other branch methods.

Part of #7347.

### CODE REVIEW

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [inline clients/python-wrapper/lakefs/branch.py:229]: Clearly old docs were wrong about on this, thanks!

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [inline clients/python-wrapper/lakefs/branch.py:247]: This is a race condition, it might not return the reverted branch.  revert_branch returns the reference I think, and we can get that commit from lakeFS rather than from head.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
**Unit** tests are failing at test_branch_revert in test_branch.py.  AFAICT that test patches the revert_branch API call; obviously you will now need to patch the get_commit API call as well.

The clue is [this line](https://github.com/treeverse/lakeFS/actions/runs/7708052601/job/21097891732?pr=7353#step:7:375).  Sorry about this: (all) mocking libraries generate the worst possible error messages, and this unit test uses patching.

File: clients/python-wrapper/lakefs/branch.py
Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py:222]: So since `str` is a subtype of `ReferenceType`, we are compatible there (any previously legal input is still legal now, plus `Reference` and `Commit` objects). 
Since we are now returning a value that was previously None (and thus not interesting), we should be good there, too. Only problem I see is with people passing `reference_id` as a keyword argument to `Branch.revert()`.

Do you think that's common? I would love to at least keep that increased flexibility in typing.

File: clients/python-wrapper/lakefs/branch.py
Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py:247]: You mean a lakefs SDK API call, right? I can do that (I'll likely just steal the code from `get_commit()`).

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [reply clients/python-wrapper/lakefs/branch.py:222]: You are of course correct, on all counts.  Let's fix everything: have an argument `reference` that is a string and alias `reference_id` to that argument.  There are [decorators that can do this](https://stackoverflow.com/a/29374525/192263), but probably just 

```suggestion
    def revert(self, _reference: ReferenceType, parent_number: Optional[int] = None, reference: Optional[ReferenceType] = None, reference_id: Optional[ReferenceType] = None) -> Commit:
      # Allow reference_id, the deprecated old argument, as a keyword arg.
      reference = _reference or reference or reference_id
```

will mostly do the trick.

@N-o-Z can we go there?  Because realistically this _is_ a mistyping, it makes a lot of sense to allow the second arg to be a reference and not just an ID, and Python handling of function args is spectacularly dangerous for software engineering.  Otherwise we break backwards compatibility in this small way, or keep a confusing arg name.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Sorry, final approval on this goes to @N-o-Z because of the arg type change.

Alternatively we can split this into 2.  Commit the return type right now, it's a backwards compatible change, and block the argument name change for some later major version.

**Comment by nicholasjng** [comment]
I'm cool with it. Does "major version" imply a new v0.x.0 here? (I think you called the v0.3.0 "version 3" if I'm not mistaken)

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [reply clients/python-wrapper/lakefs/branch.py:222]: Sure, that's a beautiful solution!  Your version will be entirely backwards compatible :-)

```suggestion
    def revert(self, reference: ReferenceType, parent_number: Optional[int] = None, *, reference_id: Optional[str] = None) -> Commit:
      # Handle reference_id as a deprecated alias to reference.
      reference = reference or reference_id
```

Comment (nicholasjng) [reply clients/python-wrapper/lakefs/branch.py]: Thanks, I took this over in [c253b34](https://github.com/treeverse/lakeFS/pull/7353/commits/c253b347777c481b70ce93ada3c88d7653e5d9b3).

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [inline clients/python-wrapper/lakefs/branch.py:251]: I think this is still a race: multiple processes may be working on the same branch.  The field `self._id` is the name of a branch, for instance `main`.  Suppose that I am running this code to revert, and concurrently Nir merges into the branch.  My code runs revert and moves `main`, now Nir merges in and moves `main` again, now my code runs get_commit.  I get Nir's merge, which is **not** the result of the merge.  That's why the revert API returns a new digest: it's the only way for the caller to know what it created!


I think that you will need something like this:
```suggestion
            new_digest = self._client.sdk_client.branches_api.revert_branch(
                self._repo_id,
                self._id,
                lakefs_sdk.RevertCreation(ref=reference_id, parent_number=parent_number)
                lakefs_sdk.RevertCreation(ref=ref, parent_number=parent_number)
            )
            commit = self._client.sdk_client.commits_api.get_commit(self._repo_id, new_digest)
```

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!  Still one last race condition, but I think we can fix it easily.

**Comment by N-o-Z** [comment]
> I'm cool with it. Does "major version" imply a new v0.x.0 here? (I think you called the v0.3.0 "version 3" if I'm not mistaken)

major == x.0.0

File: clients/python-wrapper/lakefs/branch.py
Comment (arielshaqed) [reply clients/python-wrapper/lakefs/branch.py:251]: As you rightly explained, there is a bug in our SDK: revertBranch does not return the generated commit.  So there is no way to implement this correctly on the client.  Given this, let's go with what we can do.  I shall open a bug to work around this, and future versions of client and server may be able to do this.

Please add words to the docstring such as
> ```
> Due to a limitation of lakeFS IO, the returned commit might be incorrect when
> called on the server concurrently with other operations that affect the commit
> log, such as commits, merges, cherry-picks, and reverts.
> ```

Other than that, we can pull.

THANKS!

**Comment by arielshaqed** [comment]
> > I'm cool with it. Does "major version" imply a new v0.x.0 here? (I think you called the v0.3.0 "version 3" if I'm not mistaken)
> 
> major == x.0.0

The old Semver allowed breaking changes by bumping the left-most non-zero number.   Before version 1.0.0 this means we may break backwards-compatibility on version 0.4.0.  In any case [New Semver 2](https://semver.org/#spec-item-4) allows us to do whatever we want before 1.0.0 :-/

We can do it in 0.4.0.

File: clients/python-wrapper/lakefs/branch.py
Comment (N-o-Z) [inline clients/python-wrapper/lakefs/branch.py:222]: For backwards compatibility - reference should be Optional

File: clients/python-wrapper/lakefs/branch.py
Comment (N-o-Z) [inline clients/python-wrapper/lakefs/branch.py:242]: 1. Please issue warning when `reference_id` is not None
2. Please validate `reference` is not None afterwards

File: clients/python-wrapper/tests/utests/test_branch.py
Comment (N-o-Z) [inline clients/python-wrapper/tests/utests/test_branch.py:124]: We need the following additional tests for revert_branch:
1. Pass reference_id and not reference
2. Pass reference and not reference_id
3. Don't pass any
4. Pass both

**Review by N-o-Z — CHANGES_REQUESTED** [review]
@nicholasjng Some additional comments

**Review by arielshaqed — APPROVED** [review]
Thanks!

@N-o-Z is the new pydoc what we needed?

**Comment by N-o-Z** [comment]
@nicholasjng As always thank you for the hard work!
I took the liberty to push some modifications to the code documentation as well as fix a bug in the code and the unit test.
Please take a look and let me know if I missed something. I'll merge after your approval

**Comment by nicholasjng** [comment]
That's a good change, I didn't touch it since it was already on main and not related to the reference issue. Thank you!

**Review by N-o-Z — APPROVED** [review]

**Pull request merged**
